### PR TITLE
GitHub Actions: Create Release Draft when tagging version

### PR DESCRIPTION
**Pull request by ockham**
# Description
Second stab at writing a workflow (GitHub action) to build Gutenberg upon creating a tag that starts with `v`, and creating a draft release with the build attached to it. Supersedes #19626.

The difference to #19626 is that:
- the release creation is now done in a separate job, and the plugin zip is retained as an artifact, and
- the release notes are extracted from the Changelog.

The (locally run) release script works pretty much as before -- only plugin zip creation and GitHub release creation have been removed (since they're now done by the new workflow).

Note that we're currently creating a release _draft_, meaning an extra step is required to publish the release: Clicking the 'Publish' button. While I find that extra safety net convenient, it is mainly for testing purposes. I'm happy to change it to publish the release right away -- it's only a [one-line change](https://github.com/WordPress/gutenberg/pull/27488/files#diff-40b47d855e2cffe931290bba37c1b8716642627cbe05ac859728cb95c5da7536R80) (plus a few instructions in the release scripts).

The new GH workflow uses the following GH actions:
- https://github.com/actions/create-release
- https://github.com/actions/download-artifact
- https://github.com/actions/upload-release-asset.

The main benefit of this PR is that the plugin build is decoupled from people's local systems and setups, yielding reproducible builds, and cutting down time spent running the release process locally (freeing up people's systems earlier). (One micro-optimization is e.g. that it's no longer necessary to enter a GH access token during the release process.)

## How has this been tested?

See here for an example release draft created by this action: https://github.com/ockham/gutenberg/releases/tag/untagged-f7f779e5825df118d22c

Corresponding workflow run: https://github.com/ockham/gutenberg/runs/1670519325?check_suite_focus=true

Or use the following instructions to test for yourself:

- Make sure you have your own fork of the Gutenberg repo, and that you've added it locally as a remote (E.g. `git remote add ockham git@github.com:ockham/gutenberg.git` if your GH username is `ockham` :slightly_smiling_face: ).
- Check out this PR's branch, and push the contents to your fork's `master` branch (you can reset later):
```
git push ockham add/gh-action-create-release:master
```
- Note the latest GB plugin release version (9.7.0 at the time of this writing).
- Delete the corresponding tag both locally (e.g. `git tag -d v9.7.0`) and in your fork `git push ockham :v9.7.0`.
- Recreate the tag, e.g. `git tag v9.7.0`.
- Push that tag to your fork: `git push --tags ockham`.
- Go to the 'Actions' tab of your fork's GH page (e.g. https://github.com/ockham/gutenberg/actions).
- Locate the action whose description reads "Build Gutenberg and upload to Draft Release", and click on it.
- It should take less than 5 minutes to build the plugin zip. You can watch that page as the action progresses.
- Eventually, a release draft will be created, and the plugin zip will be attached. You'll find it under 'Releases' (e.g. https://github.com/ockham/gutenberg/releases)
- Don't forget to clean up tags after running these commands, so you won't accidentally push tags that you solely created for testing to the actual Gutenberg repo later. See e.g. https://stackoverflow.com/questions/1841341/remove-local-git-tags-that-are-no-longer-on-the-remote-repository

(I've used my Gutenberg fork as an example, you can see the relevant action and release drafts from previous runs there.)

Alternatively, you can also use the local release script to create the tag and changelog. Make sure to modify `bin/plugin/config.js` to point to your forked repo!

## Types of changes
Build tools automation

### CODE REVIEW

**Comment by youknowriad** [comment]
> The (locally run) release script works pretty much as before -- only plugin zip creation and GitHub release creation have been removed (since they're now done by the new workflow).

Can we remove the remaining local step by making it a "manual workflow" (workflow_run) we could trigger from the actions page.

>  I'm happy to change it to publish the release right away

I think it makes sense to publish it.

---

So if I understand properly this is only for stable releases right,  how do we go about doing the same thing for tags like v9.6.0-rc.1  and mark the corresponding Github releases as "pre-release" ?

**Comment by ockham** [comment]
> > The (locally run) release script works pretty much as before -- only plugin zip creation and GitHub release creation have been removed (since they're now done by the new workflow).
> 
> Can we remove the remaining local step by making it a "manual workflow" (workflow_run) we could trigger from the actions page.

I'm not sure we can do that, since the local step is still required for giving the developer a chance to provide a changelog, no?
(I have a scenario in mind where we wouldn't indeed need the local script which I'll elaborate in a follow-up comment.)

> > I'm happy to change it to publish the release right away
> 
> I think it makes sense to publish it.

Okay, I'll change that :+1:

> So if I understand properly this is only for stable releases right, how do we go about doing the same thing for tags like v9.6.0-rc.1 and mark the corresponding Github releases as "pre-release" ?

The action already does that: https://github.com/WordPress/gutenberg/pull/27488/files#diff-40b47d855e2cffe931290bba37c1b8716642627cbe05ac859728cb95c5da7536R81 :sunglasses:

**Comment by ockham** [comment]
> I have a scenario in mind where we wouldn't indeed need the local script which I'll elaborate in a follow-up comment.

Some context:

The new action tightly couples release creation to tag creation (or rather, pushing a tag to the repo). This seemed to make sense to me, as we have kind of an invariant here: For every tag (that starts with a `v`), we want to create a release. Furthermore, I assume we'll continue to require creating a tag for each release. Piggybacking on the event of tag creation to me seems more streamlined/automated rather than requiring an extra step from the GUI to trigger release creation.

There are basically two reasons the local script is still required: Tag (and release branch) creation, and providing a prompt for the changelog.

The need for the latter mostly results from the fact that we keep `changelog.txt` in GitHub. If we want a newly created tag to include the changelog for that version, we have to commit it _prior_ to creating the tag. This is the major obstacle to automating the release process further, especially if we don't just want to use the output from the `changelog` commander script at face value, but want to give the developer that's in charge of the release the opportunity to edit it.

We can loosen that requirement as follows (but this is something I'd tackle in a subsequent PR, as it requires some more discussion and agreement):

We drop `changelog.txt` from GH. Instead, we tell the plugin repo upload workflow to use the release notes from the GH release as the changelog for the latest version, and to prepend it to the plugin repo's `changelog.txt`. In other words, `changelog.txt` would only live in SVN, but not GH.

---

This would allow for the following release flow, which personally I'd find the most streamlined:

My preferred end goal for the release flow is as follows:
1.  Developer creates a new tag and pushes it to GH.
2. The GH action creates a release _draft_ for it, attaching the built plugin zip as an asset, and uses the `changelog` commander script to prepopulate the release notes.
3. Developer edits the release notes, and publishes the release.
4. The plugin repo upload action picks up the release and pushes it to the plugin repo. It uses the release notes as the changelog for the new version, prepending it to the plugin repo's `changelog.txt`.

At that point, we might not actually need a local release script anymore, as all it would do would be release branch and tag creation, which are arguably easy enough to do manually.

**Comment by youknowriad** [comment]
I think at this point, I see things a bit differently, here's what would be the ideal end goal for me:

1- Go to Github actions, and choose whether to perform an RC release  or  a stable release.  (could be separate  jobs)

IF RC:

- It's an RC release,  use the changelog command to get the changelog. (and during the RC period allows devs to edit it)
- Publish the github tag, release branch (create if necessary) and create the GitHub release.

IF Stable:

- Use the latest release branch to create a release, 
- Tag the release and use the changelog from the RC release (the devs had the time to edit the changelog)
- Publish the Github release
- Publish the NPM release.

In both situations, the "release manager" only had to interact once. Click the initial job to perform the release.

--- 
In your proposal, it's not clear how the release branch gets created. It's also adds extra steps to the process. WDYT

**Comment by youknowriad** [comment]
> 1- Go to Github actions, and choose whether to perform an RC release or a stable release. (could be separate jobs)

That said, I'm fine with keeping the release script for now as the replacement of the first part of this (provide changelog, create release branch and tag) and let the action continue from there. It's still just one step to be done but locally.

**Comment by ockham** [comment]
> > 1- Go to Github actions, and choose whether to perform an RC release or a stable release. (could be separate jobs)
> 
> That said, I'm fine with keeping the release script for now as the replacement of the first part of this (provide changelog, create release branch and tag) and let the action continue from there. It's still just one step to be done but locally.

Thanks! Yeah, so the goal of the current PR was to trim down time spent on the build process locally, by moving it to the GH action. Otherwise, it doesn't really change the developer experience -- I was thinking of it as a stepping stone for future changes.

File: bin/plugin/commands/release.js [hunk 1]
```diff
@@ -650,17 +487,20 @@ async function releaseRC() {
...
-		'You can access the GitHub release here: ' +
-			formats.success( release.html_url ) +
+			formats.success( version ) +
+			' has been successfully tagged.\n',
+		"In a few minutes, you'll be able to find the GitHub release draft here: " +
+			formats.success(
+				`https://github.com/${ config.githubRepositoryOwner }/${ config.githubRepositoryName }/releases/`
+			) +
 			'\n',
+		"Don't forget to publish the release once the draft is available!\n",
```
Comment (youknowriad) [inline bin/plugin/commands/release.js:503]: Should update the wording it's an actual published release right?

Comment (ockham) [reply bin/plugin/commands/release.js:503]: Yes! Good spot, thanks!

**Comment by ockham** [comment]
> I think at this point, I see things a bit differently, here's what would be the ideal end goal for me:

Thanks for clarifying :+1: 

> 1- Go to Github actions, and choose whether to perform an RC release or a stable release. (could be separate jobs)

Yeah, they probably would need to be separate jobs :thinking: 

> IF RC:
> 
> * It's an RC release,  use the changelog command to get the changelog. (and during the RC period allows devs to edit it)

Through regular git commits?

> * Publish the github tag, release branch (create if necessary) and create the GitHub release.
> 
> IF Stable:
> 
> * Use the latest release branch to create a release,
> * Tag the release and use the changelog from the RC release (the devs had the time to edit the changelog)
> * Publish the Github release
> * Publish the NPM release.
> 
> In both situations, the "release manager" only had to interact once. Click the initial job to perform the release.
> 
> In your proposal, it's not clear how the release branch gets created.

I'd probably do that manually.

I agree that my proposal differs a bit on the philosophical front: I'd say that yours is a one-stop shop solution -- which is admittedly very convenient for release managers :smile: Mine encapsulates the GB-specific build and upload parts into actions that are tightly coupled to tag creation -- thus emphasizing the role of the latter -- and exposes the more 'generic' git parts, requiring the release manager to manually create the release branch and tags.

The reason why I prefer the latter is because it gives us more flexibility, and because I believe we cannot (nor should we) shield developers entirely from some (not very advanced) git operations. After all, AFAIU, developers will also need to potentially cherry-pick commits from `master` to the release branch for the final release. (Arguably, even modifying the changelog during the RC period through regular git commits could be seen as another example.) In other words, we have a choice where to draw the "line" (|) between where we require manual interaction, and what's automated. In the two scenarios we're discussing, it's either:

1. `git cherry-pick` | `git checkout -b`, `git tag`, build, upload, _or_,
2. `git cherry-pick`, `git checkout -b`, `git tag` | build, upload.

I find the latter preferable.

This is just the workflow we're typically dealing with now. The added flexibility of the second scenario pays off even more if we ever wish to modify that workflow. Imagine e.g. someone would want to publish a point release to a previous version :grimacing: Or any other change to the process, really.

> It's also adds extra steps to the process. WDYT

See above for the rationale about git related steps. There's still room to minimize the number of steps further, if we don't want to require the release manager to press 'Publish' on the release -- I think we can include the steps for release note creation that you describe into the current workflow.

That said, I don't think the 'optimal' workflow is solely measured in number of steps: I think it's okay to include an extra step if it
- makes sense to the person that has to perform it (and fits into their mental model, or even helps them build one), and/or
- provides an extra sense of security (e.g. being able to have a final look at a release draft prior to making it go live).

---

Happy to discuss this further so we can find a good solution that we both -- and everyone else -- likes :smile: Sorry that this turned out a bit more verbose again. (I've actually written some internal documentation on this very subject before, I'll try to find it.)

**Comment by youknowriad** [comment]
> Through regular git commits?

Just by editing the release on github

> The reason why I prefer the latter is because it gives us more flexibility, and because I believe we cannot (nor should we) shield developers entirely from some (not very advanced) git operations

I think I'm in disagreement with this because I want to be able to say anyone can do a release. And the number of times people ask on which branch they should be to do such thing or such other thing is high for me. I feel it's not really about folks not having the skills but more about removing pressure from folks to have to know exactly what they should be doing and blaming the tools if there's an error instead. Basically what I want is not saying to folks "go head these docs" to perform the release but just "go there and click there" or "just run this, it will tell you what to do". I want to remove  the stress factor and the possibility to make errors from the release manager.

But I also agree that we should ask the folks that did previous releases their opinion on this.

**Comment by ockham** [comment]
The release is now published (after attaching the asset), see https://github.com/ockham/gutenberg/runs/1681775057?check_suite_focus=true for a test run (and https://github.com/ockham/gutenberg/releases/tag/v9.7.0 for the resulting release).

I'll change the copy in the release script now.

**Comment by ockham** [comment]
> I'll change the copy in the release script now.

6bb426b

**Review by youknowriad — APPROVED** [review]
Nice

File: bin/plugin/commands/release.js (hunk 1 above)
Comment (ockham) [reply bin/plugin/commands/release.js]: Fixed in 6bb426b.

**Comment by tellthemachines** [comment]
> But I also agree that we should ask the folks that did previous releases their opinion on this.

Late to the party, but I like the idea that anyone should be able to do a release. That said, I find the hardest part of the process (assuming tooling works correctly) to be editing the changelog and writing the release post, and I suspect these tasks will be nearly impossible to anyone who isn't reasonably familiar with the repo and with ongoing work, priorities, etc.

For feedback going forward, perhaps a post on [make/core](https://make.wordpress.org/core/) would help? 

Also, let's not forget to update [the release docs](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/release.md) with the latest changes!

**Comment by ockham** [comment]
> > But I also agree that we should ask the folks that did previous releases their opinion on this.
> 
> Late to the party

Late to reply myself now -- apologies, I let this fall through the cracks.

> but I like the idea that anyone should be able to do a release. That said, I find the hardest part of the process (assuming tooling works correctly) to be editing the changelog and writing the release post, and I suspect these tasks will be nearly impossible to anyone who isn't reasonably familiar with the repo and with ongoing work, priorities, etc.

Agree, the 'editorial' part of a release is hardest.

> For feedback going forward, perhaps a post on [make/core](https://make.wordpress.org/core/) would help?

I'm planning one. Overall, once I've finished my little tool automation project (mostly #28138 at this point), I'd like to publish a post about the (hopefully) simplified release process; and to describe the role of a release manager, which would include that 'editorial' part, plus being responsible for bug fixes and patch releases over the lifecycle of a version. (This has been partly discussed with @mcsf et al).

> Also, let's not forget to update [the release docs](https://github.com/WordPress/gutenberg/blob/master/docs/contributors/release.md) with the latest changes!

I've been a bit slow with this one since I hoped to do it all in one fell swoop once I'm done with #28138, but since that has been dragging out a bit, here goes a doc update that reflects the current state: #29091.

**Pull request merged**